// angular-formio/resource/formio-resource.js
/**
 * @fileoverview added by tsickle
 * Generated from: resource/formio-resource.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { FormioResourceConfig, FormioResources, FormioResourceService, FormioResourceComponent, FormioResourceViewComponent, FormioResourceEditComponent, FormioResourceDeleteComponent, FormioResourceCreateComponent, FormioResourceIndexComponent, FormioResourceRoutes, FormioResource } from './index';
export { FormioBaseComponent as ɵf } from '../FormioBaseComponent';
export { FormioAuthConfig as ɵc } from '../auth/auth.config';
export { FormioAuthService as ɵa } from '../auth/auth.service';
export { FormioAlerts as ɵl } from '../components/alerts/formio.alerts';
export { FormioAlertsComponent as ɵj } from '../components/alerts/formio.alerts.component';
export { ParseHtmlContentPipe as ɵk } from '../components/alerts/parse-html-content.pipe';
export { FormBuilderComponent as ɵh } from '../components/formbuilder/formbuilder.component';
export { FormioComponent as ɵe } from '../components/formio/formio.component';
export { FormioLoaderComponent as ɵi } from '../components/loader/formio.loader.component';
export { CustomTagsService as ɵg } from '../custom-component/custom-tags.service';
export { FormioAppConfig as ɵb } from '../formio.config';
export { FormioModule as ɵd } from '../formio.module';
export { extendRouter as ɵx } from '../formio.utils';
export { GridBodyComponent as ɵr } from '../grid/GridBodyComponent';
export { GridFooterComponent as ɵt } from '../grid/GridFooterComponent';
export { GridHeaderComponent as ɵp } from '../grid/GridHeaderComponent';
export { FormGridBodyComponent as ɵq } from '../grid/form/FormGridBody.component';
export { FormGridFooterComponent as ɵs } from '../grid/form/FormGridFooter.component';
export { FormGridHeaderComponent as ɵo } from '../grid/form/FormGridHeader.component';
export { FormioGridComponent as ɵn } from '../grid/grid.component';
export { FormioGrid as ɵm } from '../grid/grid.module';
export { SubmissionGridBodyComponent as ɵv } from '../grid/submission/SubmissionGridBody.component';
export { SubmissionGridFooterComponent as ɵw } from '../grid/submission/SubmissionGridFooter.component';
export { SubmissionGridHeaderComponent as ɵu } from '../grid/submission/SubmissionGridHeader.component';
